### PR TITLE
feat: add MarketMap query to oracle server

### PR DESCRIPTION
**Pull request by technicallyty**
- adds ~an http handler~ gRPC handler to the Oracle server to get latest marketmap from the ProviderOrchestrator. 
- adds option to Oracle constructor to pass in a function to get a marketmap
- extends Oracle interface with `GetMarketMap` method

### CODE REVIEW

File: tests/integration/go.mod [hunk 1]
```diff
@@ -13,7 +13,7 @@ go 1.22.2
 require (
 	cosmossdk.io/math v1.3.0
 	github.com/cometbft/cometbft v0.38.7
-	github.com/cosmos/cosmos-sdk v0.50.6
+	github.com/cosmos/cosmos-sdk v0.50.7
```
Comment (technicallyty) [inline tests/integration/go.mod]: GoLand will not stop updating this.... SEND HELP

File: oracle/oracle.go [hunk 2]
```diff
@@ -259,3 +264,13 @@ func (o *OracleImpl) GetPrices() types.Prices {
...
+
+// GetMarketMap returns the current market map configuration.
+//
+// TODO(Tyler): is the empty map ok? maybe? maybe not. lord save us all.
+func (o *OracleImpl) GetMarketMap() marketmaptypes.MarketMap {
+	if o.marketMapGetter != nil {
+		return o.marketMapGetter()
+	}
+	return marketmaptypes.MarketMap{}
+}
```
Comment (technicallyty) [inline oracle/oracle.go]: if no option is passed in, ok to just empty return?

File: service/servers/oracle/server.go [hunk 3]
```diff
@@ -195,6 +196,20 @@ func (os *OracleServer) Prices(ctx context.Context, req *types.QueryPricesReques
 	}
 }
 
+func (os *OracleServer) HandleGetMarketMap(w http.ResponseWriter, _ *http.Request) {
```
Comment (g3ndo9kar9) [inline service/servers/oracle/server.go]: Instead of writing our own HTTP handler implementations, we usually define our services via [proto](https://github.com/skip-mev/slinky/blob/1657e81e45843d82d1831a18d2795d3fbae8b056/proto/slinky/service/v1/oracle.proto#L14), and use grpc's [grpc-gateway](https://github.com/grpc-ecosystem/grpc-gateway) functionality to spin up a proxy in front of our GRPC server to translate inbound REST requests. The advantage of doing this is that we can 1. define our req/res schema via protobuf, 2. version that schema better, 3. follow idioms defined in the codebase

TLDR: is that I recommend extending the service defined [here](https://github.com/skip-mev/slinky/blob/main/proto/slinky/service/v1/oracle.proto#L14) to include a MarketMap API, implement the service on the `OracleServer` as a GRPC route, and the REST route should be handled for us.

File: oracle/oracle.go (hunk 2 above)
Comment (g3ndo9kar9) [reply oracle/oracle.go]: im cool w that

Comment (g3ndo9kar9) [reply oracle/oracle.go]: Alternatively, you can pass the `getMarketMapFn` closure as an argument to construct the `Oracle`, and fail the constructor if it's nil. Either way works IMO

File: tests/integration/go.mod (hunk 1 above)
Comment (g3ndo9kar9) [reply tests/integration/go.mod]: changelog is [here](https://github.com/cosmos/cosmos-sdk/releases/tag/v0.50.7), doesn't look like anything super relevant to us, so this is safe+

**Review by g3ndo9kar9 — COMMENTED** [review]
one comment, congrats on becoming an official oracle 👨‍🍳

File: service/servers/oracle/server.go (hunk 3 above)
Comment (technicallyty) [reply service/servers/oracle/server.go]: cc @Eric-Warehime had mentioned to just make it a http handler -- can switch to gRPC if thats desired! i agree it would make it more uniform w/ rest of code 👍🏻

File: oracle/oracle.go (hunk 2 above)
Comment (technicallyty) [reply oracle/oracle.go]: totally -- just wasn't sure how delicate we wanted to be w/ breaking the constructor signature. 

or perhaps i should find a way to serve a default to the static config.. cause theres _always_ a marketmap, right? this endpoint is just to view it as it changes, so maybe serving whatever default (if there is one) would be nicer here?

File: cmd/slinky/main.go [hunk 4]
```diff
@@ -289,7 +289,11 @@ func runOracle() error {
 	defer orch.Stop()
 
 	// Create the oracle and start the oracle server.
-	oracleOpts = append(oracleOpts, oracle.WithProviders(orch.GetPriceProviders()))
+	oracleOpts = append(
+		oracleOpts,
+		oracle.WithProviders(orch.GetPriceProviders()),
+		oracle.WithMarketMapGetter(orch.GetMarketMap),
```
Comment (aljo242) [inline cmd/slinky/main.go:295]: Why are we making this optional?

I think we should just pass a reference to the `orchestrator` into the `oracle` upon construction and use the `GetMarketMap` function directly.  There should only ever be one implementation of this `MarketMapGetter` function.

Comment (aljo242) [reply cmd/slinky/main.go:295]: This is kind of a broader discussion about how we construct the oracle in general.  Right now the oracle does not "own" or have a notion of the orchestrator, which seems confusing to me

Comment (technicallyty) [reply cmd/slinky/main.go:295]: yeah, that was confusing to me as well. it seemed the oracle needs to know about things the orchestrator is doing, so i thought id find a direct ref to it but found the options instead

i can address that in this PR if we'd like or can open a new linear ticket?

Comment (davidterpay) [reply cmd/slinky/main.go:295]: yea ideally we move away from the price oracle and move core logic into the orchestrator and make it the central info maintainer. Definitely not in scope for this PR, would open a separate issue.

Comment (davidterpay) [reply cmd/slinky/main.go:295]: Given we now have the RPC we can probably also remove the flag to write the marketmap locally (can also be a separate issue) in the orchestrator.

File: service/servers/oracle/server.go (hunk 3 above)
Comment (davidterpay) [reply service/servers/oracle/server.go]: also in favor of just adding an additional RPC method in proto.

File: cmd/slinky/main.go (hunk 4 above)
Comment (aljo242) [reply cmd/slinky/main.go:295]: agreed - let's open a separate issue for refactoring the logic.


I think for the meantime, maybe we can just pass a ref to the orchestrator and always enable queries?

Comment (technicallyty) [reply cmd/slinky/main.go:295]: ticket [here](https://linear.app/skip/issue/BLO-1329/move-oracle-service-into-orchestrator), marking this as resolved

File: oracle/oracle.go (hunk 2 above)
Comment (technicallyty) [reply oracle/oracle.go]: gonna mark this as resolved as this will be a tad different after BLO-1329

File: service/servers/oracle/server.go (hunk 3 above)
Comment (technicallyty) [reply service/servers/oracle/server.go]: this is now switched to gRPC, @davidterpay @nivasan1

**Review by aljo242 — APPROVED** [review]

**Review by Eric-Warehime — APPROVED** [review]
:shipit:
